src/boosting/dart.cpp

```diff
@@ -0,0 +1,130 @@
+#include "gbdt.h"
+#include "dart.h"
+
+#include <LightGBM/utils/common.h>
+
+#include <LightGBM/feature.h>
+#include <LightGBM/objective_function.h>
+#include <LightGBM/metric.h>
+
+#include <ctime>
+
+#include <sstream>
+#include <chrono>
+#include <string>
+#include <vector>
+#include <utility>
+
+namespace LightGBM {
+
+DART::DART(){
+}
+
+DART::~DART(){
+}
+
+void DART::Init(const BoostingConfig* config, const Dataset* train_data, const ObjectiveFunction* object_function,
+     const std::vector<const Metric*>& training_metrics) {
+  GBDT::Init(config, train_data, object_function, training_metrics);
+  gbdt_config_ = dynamic_cast<const GBDTConfig*>(config);
+  drop_rate_ = gbdt_config_->drop_rate;
+  random_for_drop_ = Random(gbdt_config_->dropping_seed);
+}
+
+bool DART::TrainOneIter(const score_t* gradient, const score_t* hessian, bool is_eval) {
+  // drop trees
+  double shrinkage_rate = DroppingTrees();
```

Comment:
OK. But in this case, there is no need to use dart rather than gbdt. I forgot to say this: I think it's still doable to use self-defined gradients if the case is self-defined objective function. Because user still needs 'scores' (or called current prediction) to calculate the gradients. We can drop the trees, update the scores, pass back the scores to somewhere and calculate gradients, and do the remaining steps. I am not sure how you implement self-defined objective function now, so I will leave this now and change it with other together.